backend/pkg/database/refresh_materialized_views.go

```diff
@@ -0,0 +1,129 @@
+// Copyright © 2022 Cisco Systems, Inc. and its affiliates.
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package database
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"gorm.io/gorm"
+)
+
+const DefaultViewRefreshIntervalSecond = 5
+
+type refreshFunc func(db *gorm.DB, viewNames []string)
+
+type ViewRefreshHandler struct {
+	mu                        sync.Mutex
+	viewsToRefresh            map[string][]string // map of tables that shows which views should be refreshed
+	tableChanged              map[string]bool
+	refreshFunc               refreshFunc
+	viewRefreshIntervalSecond time.Duration
+}
+
+func (vh *ViewRefreshHandler) TableChanged(table string) {
+	vh.mu.Lock()
+	defer vh.mu.Unlock()
+	vh.tableChanged[table] = true
+}
+
+func (vh *ViewRefreshHandler) GetAndClearChanges() map[string]bool {
+	vh.mu.Lock()
+	defer vh.mu.Unlock()
+	tables := vh.tableChanged
+	vh.tableChanged = make(map[string]bool)
+	return tables
+}
+
+func (vh *ViewRefreshHandler) runRequiredRefreshes(db *gorm.DB) {
+	viewNames := vh.getViewsToRefresh()
+	vh.refreshFunc(db, viewNames)
+}
+
+func (vh *ViewRefreshHandler) getViewsToRefresh() []string {
+	viewToRefresh := make(map[string]bool)
+	tables := vh.GetAndClearChanges()
+	for table, ok := range tables {
+		if ok {
+			for _, viewName := range vh.viewsToRefresh[table] {
+				viewToRefresh[viewName] = true
+			}
+		}
+	}
+	views := make([]string, 0)
+	for viewName, ok := range viewToRefresh {
+		if ok {
+			views = append(views, viewName)
+		}
+	}
+	return views
```

Comment:
honestly I wouldn't even convert this to a list of strings when you can just loop over the set like in my previous review

Reply:
We iterate over the map of tables that were changed and based on the mapping between tables and views in this way we create a simple list that can be passed to the `refreshFunc` to iterate over it.

Reply:
you already have a simple list (a set) viewsToRefresh, you can iterate over it like:

```
for viewName, _ := viewsToRefresh {
     refreshFunc(viewName)
}
```

no need to convert its just a waste of time and memory.

Reply:
Oh, you mean only the last conversion. You're right. I will udate